File: sdks/python/fedichess/examples/lobby_and_challenge.py
#!/usr/bin/env python3
"""
Example: join lobby, send heartbeat, list peers, optionally send a challenge.
Run the bridge first (see sdks/bridge/README.md). From sdks/python:
  python -m fedichess.examples.lobby_and_challenge
"""
import os
import sys
import time
import uuid

from fedichess import FediChessClient


def main() -> None:
    # Bridge at sdks/bridge/dist/index.js (relative to this file: ../../bridge)
    this_dir = os.path.dirname(os.path.abspath(__file__))
    bridge_dir = os.path.join(this_dir, "..", "..", "..", "bridge")
    bridge_path = os.path.join(bridge_dir, "dist", "index.js")
    if not os.path.isfile(bridge_path):
        bridge_path = os.environ.get("FEDICHESS_BRIDGE", "node")
    bridge_cwd = bridge_dir if os.path.isdir(bridge_dir) else None

    client = FediChessClient(bridge_path=bridge_path, bridge_cwd=bridge_cwd)
    client.start_bridge()

    r = client.join_lobby()
    if not r.get("ok"):
        print("join_lobby failed:", r.get("error", r))
        return

    client.send("heartbeat", {
        "id": "sdk-python-example",
        "elo": 1200,
        "name": "PythonSDK",
        "ready": True,
        "timestamp": int(time.time() * 1000),
    })

    for _ in range(3):
        ev = client.poll_event(timeout=2.0)
        if ev:
            print("Event:", ev.get("event"), ev.get("peerId", ""), ev.get("payload"))

    peers = client.get_peers()
    print("Peers in lobby:", peers)

    if peers:
        game_id = str(uuid.uuid4())
        client.send("challenge", {
            "type": "challenge",
            "gameId": game_id,
            "challengerId": "sdk-python-example",
            "challengerName": "PythonSDK",
            "challengerElo": 1200,
            "color": "w",
            "timestamp": int(time.time() * 1000),
        }, peer_id=peers[0])
        print("Sent challenge to", peers[0], "gameId", game_id)
        for _ in range(10):
            ev = client.poll_event(timeout=2.0)
            if ev and ev.get("event") == "challResp":
                print("challResp:", ev.get("payload"))
                break

    client.leave_lobby()
    client.stop()
    print("Done.")


if __name__ == "__main__":
    main()

File: sdks/rust/examples/lobby_and_challenge.rs
//! Example: join lobby, send heartbeat, list peers, optionally send a challenge.
//! Run from sdks/rust: cargo run --example lobby_and_challenge
//! Build the bridge first: cd ../../bridge && npm install && npm run build

use fedichess_sdk::{Event, FediChessClient};
use std::path::Path;
use std::time::{SystemTime, UNIX_EPOCH};

fn main() -> std::io::Result<()> {
    let bridge_dir = Path::new(env!("CARGO_MANIFEST_DIR")).join("..").join("..").join("bridge");
    let bridge_path = bridge_dir.join("dist").join("index.js");
    let bridge_path_str = bridge_path.to_string_lossy();
    let (path, cwd) = if bridge_path.is_file() {
        (bridge_path_str.as_ref(), Some(bridge_dir.to_string_lossy().as_ref()))
    } else {
        (std::env::var("FEDICHESS_BRIDGE").as_deref().unwrap_or("node"), None)
    };

    let mut client = FediChessClient::new(path, cwd)?;

    let r = client.join_lobby()?;
    if !r.ok {
        eprintln!("join_lobby failed: {:?}", r.error);
        return Ok(());
    }

    let ts = SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_millis() as u64;
    client.send(
        "heartbeat",
        serde_json::json!({
            "id": "sdk-rust-example",
            "elo": 1200,
            "name": "RustSDK",
            "ready": true,
            "timestamp": ts,
        }),
        None,
    )?;

    for _ in 0..3 {
        if let Some(ev) = client.poll_event() {
            println!("Event: {} {:?} {:?}", ev.event, ev.peer_id, ev.payload);
        }
    }

    let peers = client.get_peers()?;
    println!("Peers in lobby: {:?}", peers);

    if let Some(first_peer) = peers.first() {
        let game_id = uuid::Uuid::new_v4().to_string();
        let ts = SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_millis() as u64;
        client.send(
            "challenge",
            serde_json::json!({
                "type": "challenge",
                "gameId": game_id,
                "challengerId": "sdk-rust-example",
                "challengerName": "RustSDK",
                "challengerElo": 1200,
                "color": "w",
                "timestamp": ts,
            }),
            Some(first_peer.as_str()),
        )?;
        println!("Sent challenge to {} gameId {}", first_peer, game_id);
        for _ in 0..10 {
            if let Some(ev) = client.poll_event() {
                if ev.event == "challResp" {
                    println!("challResp: {:?}", ev.payload);
                    break;
                }
            }
        }
    }

    client.leave_lobby()?;
    client.stop();
    println!("Done.");
    Ok(())
}
